engine/src/cmd/unit_csv.cpp

```diff
@@ -809,71 +809,109 @@ void Unit::LoadRow(std::string unit_identifier, string modification, bool saved_
-                    shield_values[i] = 0.0f;
+                    VS_LOG(error, (boost::format("%1%: Unable to convert shield value '%2%' to a number: %3%") % __FUNCTION__ % shield_string_value % ex.what()));
+                    shield_sections[shield_keys[i]] = 0.0F;
+                    shield_values.emplace_back(0.0F);
+                } catch (std::out_of_range const& ex) {
+                    VS_LOG(error, (boost::format("%1%: Unable to convert shield value '%2%' to a number: %3%") % __FUNCTION__ % shield_string_value % ex.what()));
+                    shield_sections[shield_keys[i]] = 0.0F;
+                    shield_values.emplace_back(0.0F);
                 }
             }
         }
 
         if (shield->number_of_facets == 4 || shield->number_of_facets == 2) {
-            shield->UpdateFacets(shield->number_of_facets, shield_values);
+            shield->UpdateFacets(shield->number_of_facets, shield_values.data());
         }
     } else {
-        // Fallback to old
+        // Fallback to old shield_keys
         int shield_count = 0;
-        float shield_values[4];
-
-        // TODO: this mapping should really go away
-        // I love macros, NOT.
-        shield_string_values[0] = UnitCSVFactory::GetVariable(unit_key, "Shield_Front_Top_Right", std::string());
-        shield_string_values[1] = UnitCSVFactory::GetVariable(unit_key, "Shield_Back_Top_Left", std::string());
-        shield_string_values[2] = UnitCSVFactory::GetVariable(unit_key, "Shield_Front_Bottom_Right", std::string());
-        shield_string_values[3] = UnitCSVFactory::GetVariable(unit_key, "Shield_Front_Bottom_Left", std::string());
-
-        for (int i = 0; i < 4; i++) {
-            shield_values[i] = 0.0f;
-
-            if (shield_string_values[i].empty()) {
-                continue;
+        std::vector<std::string> tmp_keys{"Shield_Front_Top_Right", "Shield_Back_Top_Left", "Shield_Front_Bottom_Right", "Shield_Front_Bottom_Left"};
+        std::vector<float> shield_values{};
+
+        try {
+            for (auto &key : tmp_keys) {
+                std::string tmp_string_value = UnitCSVFactory::GetVariable(unit_key, key, std::string());
+                float tmp_float_value = std::stof(tmp_string_value);
+                shield_sections[key] = tmp_float_value;
+                shield_values.emplace_back(tmp_float_value);
+                shield_string_values.emplace_back(tmp_string_value);
+                ++shield_count;
             }
-
-            shield_values[i] = ::stof(shield_string_values[i]);
-            // Should add up to the shield type - quad or dual
-            shield_count++;
-        }   
+        } catch (std::exception const& ex) {
+            shield_count = 0;
+            shield_string_values.clear();
+            shield_values.clear();
+            shield_sections.clear();
+            tmp_keys.clear();
+            tmp_keys.emplace_back("shield_front");
```

Comment:
Why not
`tmp_keys.insert(tmp_keys.end(), { "shield_front" ... });`
?

Reply:
My understanding is that `emplace_back` is one of the more efficient ways to add an element to the end of a collection. More efficient than `push_back`, for example.

Reply:
Maybe. But this is negligible. This is a non-repeating piece of code that runs when a unit is created.